napari/_vispy/_tests/test_vispy_points_layer.py

```diff
@@ -47,18 +47,18 @@ def test_change_text_updates_node_string():
 def test_change_text_color_updates_node_color():
     points = np.random.rand(3, 2)
     properties = {'class': np.array(['A', 'B', 'C'])}
-    text = {'text': 'class', 'color': [1, 0, 0]}
+    text = {'string': 'class', 'color': [1, 0, 0]}
```

Comment:
We discussed this in the weekly architecture meeting, and none of us particularly like this pattern. Specifically, some black magic is being done to determine that `string` refers to the string encoding on the text, and it is being set to 'class', which, you might note, is not a string encoding but a plain old string. However, this magically checks for 'class' in the properties table, and if it is present, sets this as the source data for the encoding. If *not*, then it is set as a constant encoding.

This type of magic variable is, at least in this case and imho, confusing to both read and reason about. Additionally, @andy-sweet pointed out that removing this magic would simplify a lot of code, so we all agreed that we would remove it. The proposed syntax would instead be something along the lines of:

```python
text = {
    'string_encoding': {'property': 'class'},
    'color_encoding': {'constant': [1, 0, 0]}
}
```

The color encoding bit I don't think we discussed specifically, and, as it is less ambiguous, I'm not set on it, but it would be nice to have everything be consistent for sure.

Reply:
+1

Reply:
Done naively, this will break the public API because we currently support this black magic in `TextManager` and also in some other places where we currently handle encoding colors from properties (e.g. `ColorManager`).

As we're deprecating the `text` init parameter in favor of the `string` or `string_encoding` field, I think we can handle this (e.g. by only doing the black magic on the `text` parameter in `TextManager.__init__` and not in the `TextManager.string` setter), but just wanted to respond as I work through this.

Reply:
Actually, this is more of a problem because `Points.text` or the `text` init parameter, can also be specified as a `str` that refers to a property name - with the intent that the strings will be directly mapped from that property column. The same thing happens for `Shapes`.

I guess we could issue a deprecation warning, but maintain that behavior for now. I worry about how this could be confusing though.

Reply:
hmm, i can see that being confusing too. maybe we deprecate the init behaviour?

Reply:
It's easy enough to do. The only problem I see is that the way to get the same behavior will become more verbose. I.e. from something like this:

```python
points = Points(data, properties=properties, text='class')
```

to something like this

```python
points = Points(data, properties=properties, text={'string': {'property': 'point_type'}})
```

Actually, now I look at it, I don't mind the verbosity too much. The text argument becomes quite self-descriptive in spite of its length.

Reply:
> Actually, now I look at it, I don't mind the verbosity too much. The text argument becomes quite self-descriptive in spite of its length.

yeah me too, i don't mind that, it's fairly easy to parse. Wasn't @jni advocating for `string_encoding` here? Or am i getting confused

Reply:
Just to summarize, I think there are two main problems here.
1. This is fairly big public API change/deprecation, because doing things like `points.text = 'class'` and `points.face_color = 'class'` are part of our examples on `main`.
2. This drifts away from Altair's API, where something like `color='class'` is the standard way of specifying visualization style that is given in examples (maybe with the type as well).

These are related. I think one way to justify the change is that our default styling is typically constant. Partly because by default I think we want to be fast and simple, rather than expressive. We also want to allow expressive ways to style data, but the user has to do a little more work to set that up.

Reply:
> Wasn't @jni advocating for `string_encoding` here?

Yes, I saw that, but I see this a bit differently. This is partly related to my comment above where I think the default behavior and API should be the most easily understood.

I.e. I think something like 

```python
points.face_color_encoding = 'red'
```

whereas I think something like

```python
points.face_color = {'property': 'class'}
```

is understandable enough because the RHS expresses that this is not simply assignment to one color.

That being said, one benefit of using the `_encoding` names is that there's a clearer path to deprecation because things like `Points.face_color` are already used.

Sorry, this is a long point on naming. I was going to bring some naming issues up in the community meeting today, so that's probably a better place to discuss.

Reply:
> Sorry, this is a long point on naming.

It's ok, unfortunately such an important thing to get right. thanks for adding additional detail